network/data_link/wifi.c

```diff
@@ -0,0 +1,433 @@
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+
+static int32_t wifi_socket_open(struct wifi_desc *desc, uint32_t *sock_id,
+				enum socket_protocol proto);
+static int32_t wifi_socket_close(struct wifi_desc *desc, uint32_t sock_id);
+static int32_t wifi_socket_connect(struct wifi_desc *desc, uint32_t sock_id,
+				   struct socket_address *addr);
+static int32_t wifi_socket_disconnect(struct wifi_desc *desc, uint32_t sock_id);
+static int32_t wifi_socket_send(struct wifi_desc *desc, uint32_t sock_id,
+				const void *data, uint32_t size);
+static int32_t wifi_socket_recv(struct wifi_desc *desc, uint32_t sock_id,
+				void *data, uint32_t size);
+static int32_t wifi_socket_sendto(struct wifi_desc *desc, uint32_t sock_id,
+				  const void *data, uint32_t size,
+				  struct socket_address to);
+static int32_t wifi_socket_recvfrom(struct wifi_desc *desc, uint32_t sock_id,
+				    void *data, uint32_t size,
+				    struct socket_address *from);
+
+/* Connect internal functions to the network interface */
+static void wifi_init_intarface(struct wifi_desc *desc)
+{
+
+	desc->interface.net = desc;
+	desc->interface.socket_open =
+		(int32_t (*)(void *, uint32_t *, enum socket_protocol))
+		wifi_socket_open;
+	desc->interface.socket_close =
+		(int32_t (*)(void *, uint32_t))
+		wifi_socket_close;
+	desc->interface.socket_connect =
+		(int32_t (*)(void *, uint32_t, struct socket_address *))
+		wifi_socket_connect;
+	desc->interface.socket_disconnect =
+		(int32_t (*)(void *, uint32_t))
+		wifi_socket_disconnect;
+	desc->interface.socket_send =
+		(int32_t (*)(void *, uint32_t, const void *, uint32_t))
+		wifi_socket_send;
+	desc->interface.socket_recv =
+		(int32_t (*)(void *, uint32_t, void *, uint32_t))
+		wifi_socket_recv;
+	desc->interface.socket_sendto =
+		(int32_t (*)(void *, uint32_t, const void *, uint32_t,
+			     const struct socket_address *))
+		wifi_socket_sendto;
+	desc->interface.socket_recvfrom =
+		(int32_t (*)(void *, uint32_t, void *, uint32_t,
+			     struct socket_address *))
+		wifi_socket_recvfrom;
+}
+
+/**
+ * @brief Allocate resources and initializes a wifi descriptor
+ * @param desc - Address where to store the wifi descriptor
+ * @param param - Initializing data
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t wifi_init(struct wifi_desc **desc, struct wifi_init_param *param)
+{
+	struct wifi_desc	*ldesc;
+	struct at_init_param	at_param;
+	int32_t			result;
+	union in_out_param	par;
+
+	if (!desc || !param)
+		return FAILURE;
+
+	ldesc = (struct wifi_desc *)calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return FAILURE;
+
+	at_param.irq_desc = param->irq_desc;
+	at_param.uart_desc = param->uart_desc;
+	at_param.uart_irq_conf = param->uart_irq_conf;
+	at_param.uart_irq_id = param->uart_irq_id;
+
+	result = at_init(&ldesc->at, &at_param);
```

Comment:
Does `at_init()` allocate memory to the `desc->at` pointer? If it does maybe the error code after this call should include a call to `at_remove` as well.